src/core/probe/probe.rs

```diff
@@ -0,0 +1,39 @@
+#![allow(dead_code)] // FIXME
+
+use std::fmt;
+
+use crate::core::kernel;
+
+/// Probe types supported by this program. This is the main object given to
+/// tracing APIs and it does contain everything needed to target a symbol in a
+/// given running program.
+pub(crate) enum Probe {
+    Kprobe(kernel::Symbol),
+    RawTracepoint(kernel::Symbol),
+}
+
+// Use mem::variant_count::<Probe>() when available in stable.
+pub(crate) const PROBE_VARIANTS: usize = 2;
+
+impl Probe {
+    /// We do use probe types as indexes, the following makes it easy.
+    ///
+    /// TODO: use #[repr(usize)] when Rust 1.66.0 is generally available.
```

Comment:
Arg! I don't think think `#[repr(usize)]` is going to help us here: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=829cf1e3779502552116a987a9feff8b

It seems that if we use a Variant with actual types (in this case `kernel::Symbol`s), the `repr(usize)` does not work as expected.
Am I missing something?

Reply:
Also, looking at `Kernel` I see two options for the implementation of Userspace probes:
A) `struct Kernel` is reused (and renamed into something like `ProbeMgr`, and distinguish between user and kernel probes internally (e.g: userspace probes don't accept dynamic probes, nor more than one hook per probe, nor config map, etc)
B) We add `struct User` with a very similar API nicely based on `Probe` object. Now, if that's the case `Probe` will include USDT but `Kernel` will not need to increase the size of its vectors...

Thoughts?

Reply:
> To be explicit, I'm suggesting to move the `Kernel` implementation one module up and keep the kernel-specific implementations inside `core::probe::kernel`.

See #62 as a proposed implementation of this